File: baekjoon/4485.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#define INF 987654321
#define endl '\n'
using namespace std;
using ull = unsigned long long;

int N; 
const int x_move[] = {-1,1,0,0}; // 상 하 좌 우
const int y_move[] = {0,0,-1,1};

vector<vector<ull>> matrix;
vector<vector<ull>> dis;

class Edge
{
public:
    int xpos;
    int ypos;
    ull cost;

    Edge(int xpos, int ypos, ull cost):
        xpos(xpos), ypos(ypos), cost(cost) {}
    
    bool operator > (const Edge& other) const
    {
        return this->cost > other.cost;
    }
};

template<typename T>
void showContainer(const T& t)
{
    for (int i = 1; i <= N; i++)
    {
        for (int j = 1; j <= N; j++)
        {
            cout << t[i][j] << " ";
        }cout << endl;
    }
}

void init()
{
    matrix.assign(N+1, vector<ull>(N+1,0));
    dis.assign(N+1, vector<ull>(N+1, INF));
}

void input()
{
    for (int i = 1; i <= N; i++)
        for (int j = 1; j <= N; j++)
            cin >> matrix[i][j];   
}

bool inTheMatrix(int xpos, int ypos)
{
    return xpos >= 1 && xpos <= N && ypos >= 1 && ypos <= N;
}

void dijkstra()
{
    priority_queue<Edge,vector<Edge>, greater<Edge>> pq;
    dis[1][1] = matrix[1][1];

    pq.push(Edge(1,1,dis[1][1]));

    while (!pq.empty())
    {
        auto edge = pq.top();
        pq.pop();

        int xpos = edge.xpos;
        int ypos = edge.ypos;
        ull cost = edge.cost;

        if (dis[xpos][ypos] < cost)
            continue;

        for (int i = 0; i < 4; i++)
        {
            int x = xpos + x_move[i];
            int y = ypos + y_move[i];

            if (!inTheMatrix(x,y))
                continue;
            

            ull tmp = dis[xpos][ypos] + matrix[x][y];

            if (dis[x][y] > tmp)
            {
                dis[x][y] = tmp;
                pq.push(Edge(x,y,dis[x][y]));
            }
        }
    }

}

void solve()
{
    int cnt = 1;
    while (true)
    {
        cin >> N;
        if(N==0)
            return ;
        
        init();
        input();
        dijkstra();
        cout << "Problem " << cnt << ": " << dis[N][N] << endl;
        ++cnt;
    }
}


int main(void)
{
    ios::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);


    solve();

    return (0);
}

File: baekjoon/java/4485.java
import java.io.*;
import java.util.*;

public class Main {
	
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static String buffer;
	static StringTokenizer st;
	
	static int Len;
	static int[][] matrix, dis;
	static int[] x_move = {-1,1,0,0}, y_move = {0,0,-1,1};

	// ✅ 새로운 클래스 정의
	static class Node implements Comparable<Node> {
		int cost;
		int x;
		int y;

		Node(int cost, int x, int y) {
			this.cost = cost;
			this.x = x;
			this.y = y;
		}

		@Override
		public int compareTo(Node other) {
			return Integer.compare(this.cost, other.cost);
		}
	}
	
	public static void main(String[] args) throws IOException {
		int Case = 1;
		
		while (true) {
			Len = Integer.parseInt(br.readLine());
			if (Len == 0) break;

			input();
			int Result = Dijkstra();
			System.out.printf("Problem %d: %d\n", Case, Result);
			++Case;
		}
	}
	
	static void input() throws IOException {
		matrix = new int[Len+1][Len+1];
		dis = new int[Len+1][Len+1];
		
		for (int i = 1; i <= Len; i++) {
			buffer = br.readLine();
			st = new StringTokenizer(buffer);
			Arrays.fill(dis[i], 987654321);
			for (int j = 1; j <= Len; j++) {
				matrix[i][j] = Integer.parseInt(st.nextToken());
			}
		}
	}
	
	static boolean inTheMatrix(int xpos, int ypos) {
		return xpos >= 1 && xpos <= Len && ypos >= 1 && ypos <= Len;
	}
	
	static int Dijkstra() {
		PriorityQueue<Node> pq = new PriorityQueue<>();
		pq.add(new Node(matrix[1][1], 1, 1));
		dis[1][1] = matrix[1][1];
		
		while (!pq.isEmpty()) {
			Node cur = pq.poll();
			int w = cur.cost;
			int xpos = cur.x;
			int ypos = cur.y;

			if (w > dis[xpos][ypos]) continue;
			if (xpos == Len && ypos == Len) return w;
			
			for (int i = 0; i < 4; i++) {
				int x = xpos + x_move[i];
				int y = ypos + y_move[i];
				
				if (!inTheMatrix(x, y)) continue;
				if (dis[x][y] > dis[xpos][ypos] + matrix[x][y]) {
					dis[x][y] = dis[xpos][ypos] + matrix[x][y];
					pq.offer(new Node(dis[x][y], x, y));
				}
			}
		}
		return -1;
	}
}
